### ISSUE TITLE
removing nodes from cockroach cluster breaks internal cockroach time series data

### ISSUE THREAD

**Issue by HeikoOnnebrink** [cockroachdb/cockroach#24274]
**BUG REPORT**

[config] file created at: 2018/03/28 05:38:40
I180328 05:38:40.682459 300 util/log/clog.go:938  [config] running on machine: api-4
I180328 05:38:40.682459 300 util/log/clog.go:938  [config] binary: CockroachDB CCL v1.1.6 (linux amd64, built 2018/03/12 17:58:05, go1.8.3)
I180328 05:38:40.682459 300 util/log/clog.go:938  [config] arguments: [/cockroach/cockroach start --locality=datacenter=dus21 --certs-dir=/cockroach --logtostderr --advertise-host=10.97.176.28]

2. Please describe the issue you observed:
I was just about to trash a 1.1.6 cluster that was build on slow CEPH disk. I decided to run some disaster test before I trashed it. The cluster is a 9 node cluster spanning 3 datacenter with 3 nodes in each datacenter. Locality is set for each node to the datacenter where it resides. By default replica count is 3 and replication should consider datacenter to drive replication so that each datacenter holds one replica.
Next I trashed all 3 nodes from first datacenter and expected that all will continue to work.
Connections with the client application to the surviving nodes 4-9 stayed alive and access to all table data was fine as expected.
When connecting with the Cockroach UI to e.g, node 4 I recognized that all metrics stopped to display up from the time where node 1-3 where trashed.
I also see 3 ranges as unavailable. So next I accessed UI / debug and check status of all nodes 1-9, check latency view and got proper details. 
When accessing Problem Ranges / All Problem ranges I ended up with hanging "loading cluster status..." . I tried several times and after enough tries I got some page displayed. Checking details of problem ranges also did work only 1 out of 20 tries and always ended with "Error loading range Error: Promise timed out after 100000 ms". 
After trying enough times I got some details and saw that 2 of the 3 ranges where /system/tsd .. (I assume this is internal time series data).
Problem was "No Raft Leader, Unavailable" Raft State "candidate" Lease State "expired" and Lease holder was one of the trashed nodes (n2).
Situation stayed like this and no new raft leader was elected and display of TSD data was not resumed.
Is TSD data managed / replicated in a different way than table data. Why did it suffer from loss of node 1-3?

Here some details of impacted ranges 

problem started with range : 579 / 1788 / 1890 ..

in between (one day later ) I have these unavailable ranges : 
129 / 340 / 547 / 575 / 579 / 1788 / 1890 (all unavailable & No Raft Leader)
418 / 551 / 1270/ 1611/ 1816 (No Lease)

129 >> /System/tsd/cr.node.sql.misc.count
340 >> /System/tsd/cr.node.txn.durations
547 >> /System/tsd/cr.node.txn.duration
575 >> /System/tsd/cr.store.gcbytesage
579 >> /Table/13/1/2017-12-10T11:25:16.6
1788 >> /System/tsd/cr.node.sql.mem...
1890>> /System/tsd/cr.store.lease.tra...

**Comment by knz** [comment]
So the time series are interesting, we'd need to check if the zone configs apply to them properly.

Can you share the zone configuration with us?

The range 579 is a bit more bothersome.

Paging in @a-robinson

**Comment by HeikoOnnebrink** [comment]
![temp](https://user-images.githubusercontent.com/19688787/38029337-1a851f36-3285-11e8-9554-637e192c7b33.jpg)

![temp](https://user-images.githubusercontent.com/19688787/38029374-32253e14-3285-11e8-9245-2e93c34767da.jpg)

![temp](https://user-images.githubusercontent.com/19688787/38029407-48c672e6-3285-11e8-8b9d-46f783575c4b.jpg)

![temp](https://user-images.githubusercontent.com/19688787/38029451-6ab9b250-3285-11e8-9ff2-baca99cb415d.jpg)

**Comment by HeikoOnnebrink** [comment]
in the past I used a zone config on DEFAULT zone to reduce the TTL from 24h to 600 sec.
This is removed now and the zone config is untouched.
All I do is to inject the locality info on startup (see node diagnostics/node screenshot above)

**Comment by HeikoOnnebrink** [comment]
root@api-4:/cockroach# ./cockroach zone get .default --certs-dir=/cockroach
# Server version: CockroachDB CCL v1.1.6 (linux amd64, built 2018/03/12 17:58:05, go1.8.3) (same version as client)
# Cluster ID: f7e86862-ab66-4e55-9efa-393c2c5e15c7
.default
range_min_bytes: 1048576
range_max_bytes: 67108864
gc:
  ttlseconds: 90000
num_replicas: 3
constraints: []

**Comment by HeikoOnnebrink** [comment]
./cockroach zone ls --certs-dir=/cockroach
Server version: CockroachDB CCL v1.1.6 (linux amd64, built 2018/03/12 17:58:05, go1.8.3) (same version as client)
Cluster ID: f7e86862-ab66-4e55-9efa-393c2c5e15c7
.default

**Comment by HeikoOnnebrink** [comment]
after several tries I managed to get details from r579
![temp](https://user-images.githubusercontent.com/19688787/38030592-8daa2fda-3288-11e8-99b6-b4debd188aa4.jpg)

![temp](https://user-images.githubusercontent.com/19688787/38030617-a10af8f2-3288-11e8-8afa-cb8dbea039eb.jpg)

Range log stays spinning as I assume that the source is also TSD data that is not available ..

**Comment by HeikoOnnebrink** [comment]
log from node 4 shows this kind of repeating log pattern 

I180328 13:06:54.285618 2371 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.96.214.113:26257: getsockopt: no route to host"; Reconnecting to {10.96.214.113:26257 <nil>}
I180328 13:06:55.443104 300 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.96.214.114:26257: getsockopt: no route to host"; Reconnecting to {10.96.214.114:26257 <nil>}
I180328 13:06:57.245767 291 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.96.214.117:26257: getsockopt: no route to host"; Reconnecting to {10.96.214.117:26257 <nil>}
I180328 13:06:58.445017 300 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.96.214.114:26257: getsockopt: no route to host"; Reconnecting to {10.96.214.114:26257 <nil>}
I180328 13:07:00.250182 291 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.96.214.117:26257: getsockopt: no route to host"; Reconnecting to {10.96.214.117:26257 <nil>}
I180328 13:07:01.446042 300 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.96.214.114:26257: getsockopt: no route to host"; Reconnecting to {10.96.214.114:26257 <nil>}
I180328 13:07:03.302509 2371 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.96.214.113:26257: getsockopt: no route to host"; Reconnecting to {10.96.214.113:26257 <nil>}
I180328 13:07:03.366815 114 gossip/gossip.go:478  [n4] gossip status (ok, 9 nodes)
gossip client (2/3 cur/max conns)
  5: 10.97.176.39:26257 (55h28m32s: infos 210476/2601631 sent/received, bytes 54671935B/429592982B sent/received)
  8: 10.98.215.16:26257 (29h51m41s: infos 173983/121737 sent/received, bytes 44297816B/45274910B sent/received)
gossip server (2/3 cur/max conns, infos 1302670/2957496 sent/received, bytes 246134293B/535050006B sent/received)
  9: 10.98.215.18:26257 (29h51m24s)
  7: 10.98.215.15:26257 (29h51m22s)
I180328 13:07:03.371791 116 server/status/runtime.go:223  [n4] runtime stats: 575 MiB RSS, 349 goroutines, 135 MiB/146 MiB/376 MiB GO alloc/idle/total, 111 MiB/143 MiB CGO alloc/total, 60.73cgo/sec, 0.04/0.01 %(u/s)time, 0.00 %gc (0x)
W180328 13:07:04.427614 178 storage/replica.go:1459  [timeSeriesMaintenance,n4,s4,r575/207:/System/tsd/cr.store.gcbytesa…] have been waiting 1m0s attempting to acquire lease
W180328 13:07:04.427908 9069713 storage/replica.go:2715  [timeSeriesMaintenance,n4,s4,r575/207:/System/tsd/cr.store.gcbytesa…] context cancellation after 2.1s of attempting command RequestLease [/System/tsd/cr.store.gcbytesage/9/10s/2017-12-02T02:00:00Z,/Min)
I180328 13:07:04.447144 300 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.96.214.114:26257: getsockopt: no route to host"; Reconnecting to {10.96.214.114:26257 <nil>}
I180328 13:07:06.314015 2371 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.96.214.113:26257: getsockopt: no route to host"; Reconnecting to {10.96.214.113:26257 <nil>}
I180328 13:07:07.449088 300 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.96.214.114:26257: getsockopt: no route to host"; Reconnecting to {10.96.214.114:26257 <nil>}
I180328 13:07:09.262928 291 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 10.96.214.117:26257: getsockopt: no route to host"; Reconnecting to {10.96.214.117:26257 <nil>}

**Comment by a-robinson** [comment]
Sorry for the delay on this @HeikoOnnebrink!

It looks pretty clear that a few of the cluster's ranges still had 2 copies in the first locality, leading to downtime when you took all of the first locality's nodes down.

Unfortunately this is a bit of a known limitation with the v1.1 release. It's fixed in the v2.0 release that's coming out next week, but we never cherry-picked it into a v1.1 patch release for fear of the far-reaching impact of the new code and potential massive data movement on upgrade. There's more context in https://github.com/cockroachdb/cockroach/issues/17599 and particularly the thread starting at https://github.com/cockroachdb/cockroach/issues/20241#issuecomment-359653595.

Do you have any questions? I know this kind of sucks, but it all works significantly better in the 2.0 release, and is avoidable in the meantime if you have nodes from at least 3 localities as members of the cluster when you run `cockroach init` on the cluster.

**Comment by HeikoOnnebrink** [comment]
@a-robinson thanks for the detailed feedback.. I guess I need to adjust my understanding of roach when it comes to replication. Expectation was in my example of a 9 node cluster with 3 DCs that locality=DC and replication factor 3 leads  to the right level of protection against DC outage. Now I see that I can't fully rely on it.. (at least in 1.1) Now I also understood that in case of loss of nodes not the database or table as a whole is unreachable but "problems" can happen on data level (assume due to the multi raft implementation working on range level). I assume this is something I need to share with our developers as this is some type of problem they did not face in the past.. They know to handle DB down, maybe table not reachable.. but being able to read from a table until an unavailable range will break the read is a scenario they only might know from database corruption scenario with a corrupted DB block, but this case is typically not handled by their application code. In our roach scenario of a distributed DB the situation is different and more likely to happen .. so the app code should be mature here.

Additionally when I have a DC loss and unavailable ranges are show in the UI I want to drill down to see which objects are impacted.. but exactly in this moment I cannot drill down to them via UI/debug/ranges as TSD data is unavailable. I think I need to learn more about the internals of roach and how to get/query relevant info from console. Maybe worse some blog from your side to educate us here..

In every case the topic has some high impact for me and I will invest some time here and test 2.0 release for different scenarios.. I am sure I will revert back to you ..

**Comment by dianasaur323** [comment]
@HeikoOnnebrink, we are actually currently working on a tool that will help you with this problem. Our current approach is to focus on the scenario where you lose the majority of nodes, and be able to recover from that. Your point about having just certain pieces / ranges of data being impacted is a second use case that we would eventually like to focus on as well. If you're interested, it would be great to set up some time to run some ideas by you and get feedback.

**Comment by HeikoOnnebrink** [comment]
@dianasaur323 for sure I am interested and will be available to add my input any time :-)

**Comment by a-robinson** [comment]
> @a-robinson thanks for the detailed feedback.. I guess I need to adjust my understanding of roach when it comes to replication. Expectation was in my example of a 9 node cluster with 3 DCs that locality=DC and replication factor 3 leads to the right level of protection against DC outage. Now I see that I can't fully rely on it.. (at least in 1.1)

To be fair, your understanding is the intended behavior of the system, but was incomplete in v1.1. From v2.0 onward, we'll take whatever proactive action necessary to ensure that data is spread as diversely as possible. I'm really sorry that the behavior in v1.1 doesn't meet expectations.

> Additionally when I have a DC loss and unavailable ranges are show in the UI I want to drill down to see which objects are impacted.. but exactly in this moment I cannot drill down to them via UI/debug/ranges as TSD data is unavailable. I think I need to learn more about the internals of roach and how to get/query relevant info from console. Maybe worse some blog from your side to educate us here..

Have you checked the `/#/reports/problemranges` page? It shows which ranges have which problems and doesn't rely on timeseries data. It's one of many debug pages linked to from `/#/debug`, which I don't know whether you've discovered yet. Some of the pages require knowledge of the system internals, but others can be useful to everyone.

**Comment by HeikoOnnebrink** [comment]
@a-robinson sure I know /debug /#/reports/problemranges. You are right it is not based on TSD, thanks for correction. Anyway I faced problems looking up data there up from the time when first DC was down and I got most of the time a hanging page with output "loading cluster status...". 
Anyway as v2.0 is out I already spinned up a fresh 9 node cluster and will look ahead and will repeat / continue my disaster tests there.
